File: app/src/main/java/com/anbillon/lunarlite/data/db/repository/AlmanacDataSource.java
package com.anbillon.lunarlite.data.db.repository;

import android.content.Context;

import com.anbillon.lunarlite.data.db.DaoHelper;
import com.anbillon.lunarlite.data.db.model.Almanac;
import com.anbillon.lunarlite.data.db.model.AlmanacDao;
import com.anbillon.lunarlite.domain.repository.AlmanacRepository;

import java.util.List;

import javax.inject.Inject;

import rx.Observable;

/**
 * Provide almanac data source from database.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
public class AlmanacDataSource implements AlmanacRepository {
	private AlmanacDao mAlmanacDao;

	@Inject
	public AlmanacDataSource(Context context) {
		mAlmanacDao = DaoHelper.getDaoSession(context).getAlmanacDao();
	}

	@Override
	public Observable<Almanac> getAlmanac(int wielding, int heavenlyEarthly) {
		return Observable.defer(() -> {
			List<Almanac> dataList = mAlmanacDao.queryBuilder()
				.where(AlmanacDao.Properties.Wielding.eq(wielding))
				.where(AlmanacDao.Properties.Hseb.eq(heavenlyEarthly))
				.list();

			return Observable.just(dataList.isEmpty() ? null : dataList.get(0));
		});
	}
}

File: app/src/main/java/com/anbillon/lunarlite/data/db/repository/AlmanacDataSource.kt
package com.anbillon.lunarlite.data.db.repository

import com.anbillon.lunarlite.data.db.model.Almanac
import com.anbillon.lunarlite.data.db.model.AlmanacModel
import com.coolerfall.lunarlite.domain.repository.AlmanacRepository
import com.squareup.sqlbrite2.BriteDatabase
import io.reactivex.Observable
import javax.inject.Inject

/**
 * Provide almanac data source from database.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
class AlmanacDataSource @Inject constructor(private val briteDatabase: BriteDatabase) :
    AlmanacRepository {

  override fun loadAlmanac(wielding: Int, heavenlyEarthly: Int): Observable<Almanac> {
    return briteDatabase.createQuery(AlmanacModel.TABLE_NAME,
        Almanac.selectByWieldingAndHseb(wielding,
            heavenlyEarthly)).mapToOne { Almanac.MAPPER.map(it) }
  }
}
